service_container/service_closures.rst

```diff
@@ -0,0 +1,88 @@
+.. index::
+    single: DependencyInjection; Service Closures
+
+Service Closures
+================
+
+This feature wraps the injected service into a closure allowing it to be
+lazily loaded when and if needed.
+This is useful if the service being injected is a bit heavy to instantiate
+or is used only in certain cases.
+The service is instantiated the first time the closure is called, while
+all subsequent calls return the same instance, unless the service is
+:doc:`not shared </service_container/shared>`::
+
+    // src/Service/MyService.php
+    namespace App\Service;
+
+    use Symfony\Component\Mailer\MailerInterface;
+
+    class MyService
+    {
+        /**
+         * @var \Closure
+         */
+        private $mailer;
+
+        public function __construct(\Closure $mailer)
+        {
+            $this->mailer = $mailer;
+        }
+
+        public function doSomething(): void
+        {
+            // ...
+
+            $this->getMailer()->send($email);
+        }
+
+        private function getMailer(): MailerInterface
+        {
+            return ($this->mailer)();
+        }
+    }
+
+To define a service closure and inject it to another service, create an
+argument of type ``service_closure``:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/services.yaml
+        services:
+            App\Service\MyService:
+                arguments: [!service_closure '@mailer']
+
+    .. code-block:: xml
+
+        <!-- config/services.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+
+                <service id="App\Service\MyService">
+                    <argument type="service_closure" id="mailer"/>
+                </service>
+
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        // config/services.php
+        namespace Symfony\Component\DependencyInjection\Loader\Configurator;
+
+        use App\Service\MyService;
+        use Symfony\Component\DependencyInjection\Argument\ServiceClosureArgument;
+        use Symfony\Component\DependencyInjection\Reference;
+
+        return function (ContainerConfigurator $configurator) {
+            $services = $configurator->services();
+
+            $services->set(MyService::class)
+                ->args([new ServiceClosureArgument(new Reference('mailer'))]);
+        };
```

Comment:
What about adding a `seealso` directive here pointing to service subscribers to be consistent?

Reply:
Good idea, done.